feature-account-impl/src/main/java/io/novafoundation/nova/feature_account_impl/data/fee/utils/HydraDxQuoteSharedComputation.kt

```diff
@@ -0,0 +1,76 @@
+package io.novafoundation.nova.feature_account_impl.data.fee.utils
+
+import io.novafoundation.nova.common.data.memory.ComputationalCache
+import io.novafoundation.nova.common.utils.graph.Graph
+import io.novafoundation.nova.feature_swap_core.data.assetExchange.conversion.AssetConversion
+import io.novafoundation.nova.feature_swap_core.data.assetExchange.conversion.AssetExchangeQuote
+import io.novafoundation.nova.feature_swap_core.data.assetExchange.conversion.AssetExchangeQuoteArgs
+import io.novafoundation.nova.feature_swap_core.data.assetExchange.conversion.types.hydra.HydraDxAssetConversionFactory
+import io.novafoundation.nova.feature_swap_core.data.assetExchange.conversion.types.hydra.HydraDxSwapEdge
+import io.novafoundation.nova.feature_swap_core.domain.model.SwapDirection
+import io.novafoundation.nova.runtime.ethereum.StorageSharedRequestsBuilderFactory
+import io.novafoundation.nova.runtime.multiNetwork.chain.model.Chain
+import io.novafoundation.nova.runtime.multiNetwork.chain.model.FullChainAssetId
+import io.novasama.substrate_sdk_android.extensions.toHexString
+import io.novasama.substrate_sdk_android.runtime.AccountId
+import java.math.BigInteger
+import kotlinx.coroutines.CoroutineScope
+
+class HydraDxQuoteSharedComputation(
+    private val computationalCache: ComputationalCache,
+    private val assetConversionFactory: HydraDxAssetConversionFactory,
+    private val storageSharedRequestsBuilderFactory: StorageSharedRequestsBuilderFactory
+) {
+
+    suspend fun directions(
+        chain: Chain,
+        accountId: AccountId,
+        scope: CoroutineScope
+    ): Graph<FullChainAssetId, HydraDxSwapEdge> {
+        val key = "HydraDxDirections:${accountId.toHexString()}"
+
+        return computationalCache.useCache(key, scope) {
+            val assetConversion = getAssetConversion(chain, accountId, scope)
+            assetConversion.availableSwapDirections()
+        }
+    }
+
+    suspend fun quote(
+        chain: Chain,
+        accountId: AccountId,
+        fromAsset: Chain.Asset,
+        toAsset: Chain.Asset,
+        scope: CoroutineScope
+    ): AssetExchangeQuote {
+        val key = "HydraDxQuote:${accountId.toHexString()}"
+
+        return computationalCache.useCache(key, scope) {
```

Comment:
This  is the tricky one - quotes may change based on onchain state so we should not cache `assetConversion.quote(paths, args)` if we want to be accurate